finality-aleph/src/validator_network/manager/direction.rs

```diff
@@ -0,0 +1,219 @@
+use std::collections::{HashMap, HashSet};
+
+use aleph_primitives::AuthorityId;
+
+use crate::validator_network::Data;
+
+/// Data about peers we know and whether we should connect to them or they to us. For the former
+/// case also keeps the peers' addresses.
+pub struct DirectedPeers<A: Data> {
+    own_id: AuthorityId,
+    outgoing: HashMap<AuthorityId, Vec<A>>,
+    incoming: HashSet<AuthorityId>,
+}
+
+fn bit_xor_sum_parity((a, b): (u8, u8)) -> u8 {
+    let mut result = 0;
+    for i in 0..8 {
+        result += ((a >> i) ^ (b >> i)) % 2;
+    }
+    result % 2
+}
+
+/// Whether we should call the remote or the other way around. We xor the peer ids and based on the
+/// parity of the sum of bits of the result decide whether the caller should be the smaller or
+/// greated lexicographically. They are never equal, because cryptography.
+fn should_we_call(own_id: &[u8], remote_id: &[u8]) -> bool {
+    let xor_sum_parity: u8 = own_id
+        .iter()
+        .cloned()
+        .zip(remote_id.iter().cloned())
+        .map(bit_xor_sum_parity)
+        .fold(0u8, |a, b| (a + b) % 2);
```

Comment:
what about
```rust
(x.iter().fold(0, BitXor::bitxor) ^ y.iter().fold(0, BitXor::bitxor)).count_ones() as u8 % 2
```
?

Reply:
+ you don't need `bit_xor_sum_parity`

Reply:
+ ratio + L + even better than what you put above.